src/components/TorrentDetail/Overview.vue

```diff
@@ -45,14 +47,26 @@ const shouldRefreshPieceState = computed(() => shouldRenderPieceState.value && t
  * Source:
  * https://github.com/qbittorrent/qBittorrent/blob/6229b817300344759139d2fedbd59651065a561d/src/webui/www/private/scripts/prop-general.js#L230
  */
+let piecesApp: Application | null = null
+let piecesAppLastGraphics: Graphics | null = null
 async function renderTorrentPieceStates() {
   if (!canvas.value) return
 
   const pieces = await maindataStore.fetchPieceState(props.torrent.hash)
 
-  canvas.value.width = pieces.length || -1
-  const ctx = canvas.value.getContext('2d') as CanvasRenderingContext2D
-  ctx.clearRect(0, 0, canvas.value.width, canvas.value.height)
+  // Build lookup for piece ranges of files that aren't DO_NOT_DOWNLOAD
+  // allows look up by piece index in O(log(n)) time, previous method lookup time grew quadratically or worse with number of pieces
+  const piecesSelectedRanges = new IntervalTree()
+  for (const file of cachedFiles.value) if (file.priority !== FilePriority.DO_NOT_DOWNLOAD) piecesSelectedRanges.insert([...file.piece_range], file)
+
+  canvas.value.width = 4096
+  if (piecesApp === null) {
+    piecesApp = new Application()
+    await piecesApp.init({ antialias: true, width: canvas.value.width, height: canvas.value.height })
+    ;[...canvas.value.attributes].forEach(attr => (piecesApp !== null && attr.nodeValue !== null ? piecesApp.canvas.setAttribute(attr.nodeName, attr.nodeValue) : null))
+    canvas.value.replaceWith(piecesApp.canvas)
+  }
```

Comment:
I would instantiate those in the onMounted hook instead as it only depends on the canvas being loaded in DOM, apart from that LGTM

Reply:
> I would instantiate those in the onMounted hook instead as it only depends on the canvas being loaded in DOM, apart from that LGTM

Done